pkg/query/storeset.go

```diff
@@ -229,30 +232,40 @@ func (s *StoreSet) Update(ctx context.Context) {
 		level.Warn(s.logger).Log("msg", "update of some store nodes failed", "err", err)
 	}
 
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-
-	// Record the number of occurrences of external label combinations.
+	// Record the number of occurrences of external label combinations for current store slice.
 	externalLabelStores := map[string]int{}
-	for _, st := range s.stores {
+	for _, st := range stores {
 		externalLabelStores[externalLabelsFromStore(st)]++
 	}
 
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
 	// Remove stores that where not updated in this update, because were not in s.peerAddr() this time.
 	for addr, st := range s.stores {
 		if _, ok := stores[addr]; ok {
-			if externalLabelStores[externalLabelsFromStore(st)] == 1 {
-				continue
-			}
-
-			level.Warn(s.logger).Log("msg", "dropping store, external labels are not unique", "address", st.addr)
+			continue
 		}
 
 		// Peer does not exists anymore.
 		st.close()
 	}
 
-	s.stores = stores
+	// Swap old store set with new one, excluding these with duplicated external labels.
+	s.stores = make(map[string]*storeRef, len(s.stores))
+	for addr, st := range stores {
+		// Check if it has some ext labels specified. If no, it means that it might be a store node and it is fine to have
```

Comment:
I think it's worth noting that sidecars error out directly without external labels configured, otherwise this reads like empty external set is just a heuristic.

Reply:
yea